### PR TITLE
Step logs pagination

### PR DESCRIPTION
**Pull request by schustmi**
## Describe changes
This PR adds `offset` and `length` parameters to the step logs endpoint to avoid loading the entire file in memory.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).
- [ ] If my changes require changes to the dashboard, these changes are communicated/requested.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

**Comment by strickvl** [comment]
another FYI @avishniakov since you're working on logging things at the moment

File: src/zenml/artifacts/utils.py [hunk 1]
```diff
@@ -763,7 +767,19 @@ def _load_file_from_artifact_store(
...
+                try:
+                    text_file.seek(offset, os.SEEK_END)
+                except OSError:
+                    # The negative offset was too large for the file, we seek
+                    # to the start of the file
+                    text_file.seek(0, os.SEEK_SET)
+            elif offset > 0:
+                text_file.seek(offset, os.SEEK_SET)
+
+            return text_file.read(length)
```
Comment (stefannica) [inline src/zenml/artifacts/utils.py:782]: This method is also used to load visualization files in the dashboard. You should test that still works properly with this (i.e. not sure what happens if length is None here)

File: src/zenml/zen_server/routers/steps_endpoints.py [hunk 2]
```diff
@@ -265,6 +269,10 @@ def get_step_logs(
...
     return str(
         _load_file_from_artifact_store(
-            logs.uri, artifact_store=artifact_store, mode="r"
-        )
+            logs.uri,
+            artifact_store=artifact_store,
+            mode="rb",
+            offset=offset,
+            length=length,
+        ).decode()
```
Comment (stefannica) [inline src/zenml/zen_server/routers/steps_endpoints.py:277]: why did you add the extra `decode` here ?

**Review by stefannica — APPROVED** [review]
This looks like a good quick solution for the unbound logs loading problems we've been having. It also requires some UI support to ensure proper pagination but also some buffering and pre-caching (e.g. you wouldn't want to call this endpoint for every line of log being scrolled).

File: src/zenml/zen_server/routers/steps_endpoints.py (hunk 2 above)
Comment (schustmi) [reply src/zenml/zen_server/routers/steps_endpoints.py:277]: I had to change the `mode` to binary reads, as the seeking is only available for the cloud filesystems when using this mode.

File: src/zenml/artifacts/utils.py (hunk 1 above)
Comment (schustmi) [reply src/zenml/artifacts/utils.py:782]: `read(None)` should be equivalent to just `read()` which was the previous behaviour

**Review by avishniakov — APPROVED** [review]
These changes shall land quite smoothly on the merged logs file, while the performance of multiple files logging might be not improved.
